common_crawler/main.py

```diff
@@ -101,6 +105,75 @@ def main():
         print(f"Error while saving cache manager: {e}")
 
 
+def strip_url(url: str) -> str:
+    """Strips http(s)://www. from the beginning of a url if applicable. 
+
+    Args:
+        url (str): The URL to strip.
+
+    Returns:
+        str: The stripped URL.
+    """    
+    result = re.search(r"^(?:https?://)?(?:www\.)?(.*)$", url).group(1)
+    return result
+
+
+def remove_local_duplicates(url_results: list[str]) -> list[str]:
+    """Removes duplicate URLs from a list, ignoring http(s)://www.
+
+    Args:
+        url_results (list[str]): List of URLs to deduplicate.
+
+    Returns:
+        list[str]: List of unique URLs.
+    """    
+    stripped_url_results = [strip_url(url) for url in url_results]
+    unique_urls = collections.deque()
+    adjust = 0
+
+    for index, url in enumerate(stripped_url_results):
+        if url in unique_urls:
+            del url_results[index - adjust]
+            adjust += 1
+        else:
+            unique_urls.appendleft(url)
+
+    return url_results
+
+
+def remove_remote_duplicates(url_results: list[str]) -> list[str]:
+    """Removes URLs from a list that are already present in the Label Studio project, ignoring http(s)://www.
+
+    Args:
+        url_results (list[str]): List of URLs to deduplicate.
+
+    Returns:
+        list[str]: List of remaining URLs not present in the Label Studio project.
+    """    
+    # Retrieve URLs from remote Label Studio Project
+    config = LabelStudioConfig()
+    api_manager = LabelStudioAPIManager(config)
+    response = api_manager.export_tasks_from_project(all_tasks=True)
+    remote_results = response.json()
+    
+    try:
+        remote_urls = [strip_url(result["data"]["url"]) for result in remote_results]
+    except TypeError:
+        print("Invalid Label Studio credentials. Database could not be checked for duplicates.")
+        return url_results
+    remote_urls = set(remote_urls)
+
+    stripped_url_results = [strip_url(url) for url in url_results]
+    adjust = 0
+
+    for index, url in enumerate(stripped_url_results):
+        if url in remote_urls:
+            del url_results[index - adjust]
+            adjust += 1
+
```

Comment:
This low level code should be extracted to a separate function whose name clearly indicates its functionality and with accompanying docstring and type hinting for additional clarity. Having a function that makes high level calls (as earlier in this function) and low-level calls (as here) created confusion.

Reply:
Additionally, it seems to _mostly_ follow the same low-level code in `remove_local_duplicates`: Both code can probably be extracted to a singular function which can be called in both cases (perhaps with a single parameter to govern where the logic differs, if necessary).

Reply:
Just a quick clarification on this: how much should go into a separate function? Is it the whole for loop or just the if statement or just the del statement? I haven't really hear the term 'high level calls' vs 'low level calls' before so just figured I should ask